File: command/todo.py
# ============================================
#  TODO.PY — Command Line Todo Manager
#  Language : Python 3.12+
#  Run       : python todo.py
#  Theme     : Green + Yellow — clean and human
# ============================================

from __future__ import annotations
import os
import sys
from dataclasses import dataclass, field
from typing import Optional
from enum import Enum, auto

# ── ANSI Color Constants ─────────────────────
class C:
    GREEN   = "\033[32m"
    GREEN_B = "\033[92m"
    YELLOW  = "\033[33m"
    YELLOW_B= "\033[93m"
    WHITE   = "\033[97m"
    DIM     = "\033[2m"
    BOLD    = "\033[1m"
    RED     = "\033[31m"
    RESET   = "\033[0m"
    CLEAR   = "\033[2J\033[H"

# ── App Metadata ─────────────────────────────
VERSION   = "1.0.0"
MAX_TASKS = 64

# ============================================
#  STATUS ENUM
# ============================================

class Status(Enum):
    PENDING = auto()
    DONE    = auto()

    def __str__(self) -> str:
        return "done" if self == Status.DONE else "pending"

# ============================================
#  TASK DATACLASS
# ============================================

@dataclass
class Task:
    id:     int
    title:  str
    status: Status = Status.PENDING

    @property
    def is_done(self) -> bool:
        return self.status == Status.DONE

    def complete(self) -> None:
        self.status = Status.DONE

# ============================================
#  TODO LIST
# ============================================

@dataclass
class TodoList:
    tasks:   list[Task] = field(default_factory=list)
    _next_id: int       = field(default=1, repr=False)

    # ── Add ──────────────────────────────────
    def add(self, title: str) -> Task:
        if len(self.tasks) >= MAX_TASKS:
            raise ValueError(f"Task limit reached ({MAX_TASKS}).")
        title = title.strip()
        if not title:
            raise ValueError("Title cannot be empty.")
        task = Task(id=self._next_id, title=title)
        self.tasks.append(task)
        self._next_id += 1
        return task

    # ── Find ─────────────────────────────────
    def find(self, task_id: int) -> Optional[Task]:
        return next((t for t in self.tasks if t.id == task_id), None)

    # ── Complete ─────────────────────────────
    def complete(self, task_id: int) -> Task:
        task = self.find(task_id)
        if task is None:
            raise LookupError(f"Task [{task_id}] not found.")
        if task.is_done:
            raise ValueError(f"Task [{task_id}] already completed.")
        task.complete()
        return task

    # ── Delete ───────────────────────────────
    def delete(self, task_id: int) -> Task:
        task = self.find(task_id)
        if task is None:
            raise LookupError(f"Task [{task_id}] not found.")
        self.tasks.remove(task)
        return task

    # ── Clear Completed ──────────────────────
    def clear_done(self) -> int:
        before      = len(self.tasks)
        self.tasks  = [t for t in self.tasks if not t.is_done]
        return before - len(self.tasks)

    # ── Stats ────────────────────────────────
    @property
    def done_count(self) -> int:
        return sum(1 for t in self.tasks if t.is_done)

    @property
    def pending_count(self) -> int:
        return len(self.tasks) - self.done_count

# ============================================
#  RENDERING
# ============================================

def clear_screen() -> None:
    print(C.CLEAR, end="", flush=True)

def rule(char: str = "─", width: int = 52) -> None:
    print(f"{C.DIM}{char * width}{C.RESET}")

def header() -> None:
    clear_screen()
    rule("═")
    print(
        f"{C.BOLD}{C.GREEN}"
        "  ████████╗ ██████╗ ██████╗  ██████╗ \n"
        "     ██╔══╝██╔═══██╗██╔══██╗██╔═══██╗\n"
        "     ██║   ██║   ██║██║  ██║██║   ██║\n"
        "     ██║   ██║   ██║██║  ██║██║   ██║\n"
        "     ██║   ╚██████╔╝██████╔╝╚██████╔╝\n"
        "     ╚═╝    ╚═════╝ ╚═════╝  ╚═════╝ "
        f"{C.RESET}"
    )
    print(
        f"{C.YELLOW}  CLI Task Manager  "
        f"{C.DIM}v{VERSION} — written in Python{C.RESET}"
    )
    rule("═")
    print()

def render_tasks(todo: TodoList) -> None:
    print(f"{C.BOLD}{C.WHITE}  TASKS{C.RESET}")
    rule("·")

    if not todo.tasks:
        print(f"{C.DIM}  No tasks yet. Add one below.{C.RESET}")
    else:
        for task in todo.tasks:
            if task.is_done:
                print(
                    f"{C.DIM}  [{task.id:>2}]  "
                    f"✓  {task.title}{C.RESET}"
                )
            else:
                print(
                    f"{C.GREEN}  [{task.id:>2}]{C.RESET}  "
                    f"{C.YELLOW_B}○{C.RESET}  "
                    f"{C.WHITE}{task.title}{C.RESET}"
                )

    rule("·")
    print(
        f"{C.DIM}  {todo.done_count} of "
        f"{len(todo.tasks)} completed  ·  "
        f"{todo.pending_count} remaining{C.RESET}"
    )
    print()

def render_menu() -> None:
    print(f"{C.BOLD}{C.WHITE}  ACTIONS{C.RESET}")
    rule("·")
    actions = [
        ("1", "Add task"),
        ("2", "Complete task"),
        ("3", "Delete task"),
        ("4", "Clear completed"),
        ("0", "Exit"),
    ]
    for key, label in actions:
        dim = C.DIM if key == "0" else ""
        print(
            f"{C.GREEN}  [{key}]{C.RESET}  "
            f"{dim}{C.YELLOW}{label}{C.RESET}"
        )
    rule("·")
    print(
        f"{C.GREEN_B}  › {C.RESET}",
        end="",
        flush=True
    )

# ============================================
#  INPUT HELPERS
# ============================================

def prompt(message: str) -> str:
    print(f"{C.YELLOW_B}  {message} {C.RESET}", end="", flush=True)
    try:
        return input().strip()
    except (EOFError, KeyboardInterrupt):
        return ""

def prompt_id(message: str) -> Optional[int]:
    raw = prompt(message)
    try:
        return int(raw)
    except ValueError:
        return None

def pause() -> None:
    prompt(f"{C.DIM}Press Enter to continue...{C.RESET}")

def ok(message: str) -> None:
    print(f"{C.GREEN_B}  ✓ {C.RESET}{message}")

def err(message: str) -> None:
    print(f"{C.RED}  ✗ {message}{C.RESET}")

# ============================================
#  ACTIONS
# ============================================

def action_add(todo: TodoList) -> None:
    title = prompt("New task:")
    try:
        task = todo.add(title)
        ok(f"Task [{task.id}] added.")
    except ValueError as e:
        err(str(e))

def action_complete(todo: TodoList) -> None:
    task_id = prompt_id("Mark complete — Task ID:")
    if task_id is None:
        err("Invalid ID.")
        return
    try:
        task = todo.complete(task_id)
        ok(f"Task [{task.id}] marked complete.")
    except (LookupError, ValueError) as e:
        err(str(e))

def action_delete(todo: TodoList) -> None:
    task_id = prompt_id("Delete — Task ID:")
    if task_id is None:
        err("Invalid ID.")
        return
    try:
        task = todo.delete(task_id)
        ok(f"Task [{task.id}] deleted.")
    except LookupError as e:
        err(str(e))

def action_clear(todo: TodoList) -> None:
    removed = todo.clear_done()
    if removed == 0:
        print(f"{C.DIM}  No completed tasks to clear.{C.RESET}")
    else:
        ok(f"Cleared {removed} completed task(s).")

# ============================================
#  MAIN LOOP
# ============================================

ACTIONS: dict[str, callable] = {
    "1": action_add,
    "2": action_complete,
    "3": action_delete,
    "4": action_clear,
}

def main() -> None:
    todo = TodoList()

    while True:
        header()
        render_tasks(todo)
        render_menu()

        try:
            choice = input().strip()
        except (EOFError, KeyboardInterrupt):
            break

        if choice == "0":
            clear_screen()
            print(f"{C.BOLD}{C.GREEN}  Goodbye.\n{C.RESET}")
            sys.exit(0)

        if choice in ACTIONS:
            print()
            ACTIONS[choice](todo)
            print()
            pause()
        else:
            print(f"{C.DIM}  Unknown option.{C.RESET}")
            pause()

# ============================================
#  ENTRY POINT
# ============================================

if __name__ == "__main__":
    main()

File: command/todo.rs
// ============================================
//  TODO.RS — Command Line Todo Manager
//  Language : Rust (Edition 2021)
//  Compiler : rustc
//  Run       : rustc todo.rs -o todo && ./todo
//  Theme     : Cyan + White — engineered precision
// ============================================

use std::io::{self, Write};
use std::fmt;

// ── ANSI Color Constants ──────────────────────
const CYAN:    &str = "\x1b[36m";
const CYAN_B:  &str = "\x1b[96m";
const WHITE:   &str = "\x1b[97m";
const DIM:     &str = "\x1b[2m";
const BOLD:    &str = "\x1b[1m";
const GREEN:   &str = "\x1b[32m";
const RED:     &str = "\x1b[31m";
const RESET:   &str = "\x1b[0m";
const CLEAR:   &str = "\x1b[2J\x1b[H";

// ── App Metadata ─────────────────────────────
const VERSION: &str  = "1.0.0";
const MAX_TASKS: usize = 64;

// ============================================
//  TASK STATUS
// ============================================

#[derive(Debug, Clone, PartialEq)]
enum Status {
    Pending,
    Done,
}

impl fmt::Display for Status {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Status::Pending => write!(f, "pending"),
            Status::Done    => write!(f, "done"),
        }
    }
}

// ============================================
//  TASK
// ============================================

#[derive(Debug, Clone)]
struct Task {
    id:     usize,
    title:  String,
    status: Status,
}

impl Task {
    fn new(id: usize, title: String) -> Self {
        Task {
            id,
            title,
            status: Status::Pending,
        }
    }

    fn is_done(&self) -> bool {
        self.status == Status::Done
    }
}

// ============================================
//  TODO LIST
// ============================================

struct TodoList {
    tasks:   Vec<Task>,
    next_id: usize,
}

impl TodoList {
    fn new() -> Self {
        TodoList {
            tasks:   Vec::new(),
            next_id: 1,
        }
    }

    fn add(&mut self, title: String) -> Result<usize, &'static str> {
        if self.tasks.len() >= MAX_TASKS {
            return Err("Task limit reached.");
        }
        if title.trim().is_empty() {
            return Err("Title cannot be empty.");
        }
        let id = self.next_id;
        self.tasks.push(Task::new(id, title.trim().to_string()));
        self.next_id += 1;
        Ok(id)
    }

    fn complete(&mut self, id: usize) -> Result<(), &'static str> {
        match self.tasks.iter_mut().find(|t| t.id == id) {
            Some(task) if task.is_done() => Err("Already completed."),
            Some(task) => {
                task.status = Status::Done;
                Ok(())
            }
            None => Err("Task not found."),
        }
    }

    fn delete(&mut self, id: usize) -> Result<(), &'static str> {
        let pos = self.tasks.iter().position(|t| t.id == id);
        match pos {
            Some(i) => {
                self.tasks.remove(i);
                Ok(())
            }
            None => Err("Task not found."),
        }
    }

    fn clear_done(&mut self) -> usize {
        let before = self.tasks.len();
        self.tasks.retain(|t| !t.is_done());
        before - self.tasks.len()
    }

    fn count_done(&self) -> usize {
        self.tasks.iter().filter(|t| t.is_done()).count()
    }
}

// ============================================
//  RENDERING
// ============================================

fn clear_screen() {
    print!("{}", CLEAR);
    let _ = io::stdout().flush();
}

fn print_rule(ch: char, len: usize) {
    println!("{}{}{}", DIM, ch.to_string().repeat(len), RESET);
}

fn print_header() {
    clear_screen();
    print_rule('─', 52);
    println!(
        "{}{}\
  ████████╗ ██████╗ ██████╗  ██████╗ \n\
     ██╔══╝██╔═══██╗██╔══██╗██╔═══██╗\n\
     ██║   ██║   ██║██║  ██║██║   ██║\n\
     ██║   ██║   ██║██║  ██║██║   ██║\n\
     ██║   ╚██████╔╝██████╔╝╚██████╔╝\n\
     ╚═╝    ╚═════╝ ╚═════╝  ╚═════╝ {}{}",
        BOLD, CYAN, RESET, RESET
    );
    println!(
        "{}  CLI Task Manager  {}v{} — written in Rust{}",
        CYAN_B, DIM, VERSION, RESET
    );
    print_rule('─', 52);
    println!();
}

fn render_tasks(list: &TodoList) {
    println!("{}{}  TASKS{}", BOLD, WHITE, RESET);
    print_rule('·', 52);

    if list.tasks.is_empty() {
        println!("{}  No tasks yet. Add one below.{}", DIM, RESET);
    } else {
        for task in &list.tasks {
            match task.status {
                Status::Done => {
                    println!(
                        "{}  [{:>2}]  {}✓{}  {}{}",
                        DIM, task.id,
                        GREEN, DIM,
                        task.title,
                        RESET
                    );
                }
                Status::Pending => {
                    println!(
                        "{}  [{:>2}]{}  {}○{}  {}{}{}",
                        CYAN, task.id, RESET,
                        CYAN_B, RESET,
                        WHITE, task.title, RESET
                    );
                }
            }
        }
    }

    print_rule('·', 52);
    println!(
        "{}  {} of {} completed{}",
        DIM,
        list.count_done(),
        list.tasks.len(),
        RESET
    );
    println!();
}

fn render_menu() {
    println!("{}{}  ACTIONS{}", BOLD, WHITE, RESET);
    print_rule('·', 52);
    println!("{}  [1]{}  Add task",       CYAN, RESET);
    println!("{}  [2]{}  Complete task",  CYAN, RESET);
    println!("{}  [3]{}  Delete task",    CYAN, RESET);
    println!("{}  [4]{}  Clear completed",CYAN, RESET);
    println!("{}  [0]{}{}  Exit{}",       CYAN, RESET, DIM, RESET);
    print_rule('·', 52);
    print!("{}  › {}", CYAN_B, RESET);
    let _ = io::stdout().flush();
}

// ============================================
//  INPUT HELPERS
// ============================================

fn read_line() -> String {
    let mut buf = String::new();
    io::stdin().read_line(&mut buf).unwrap_or(0);
    buf.trim().to_string()
}

fn read_id(prompt: &str) -> Option<usize> {
    print!("{}  {} {}", CYAN_B, prompt, RESET);
    let _ = io::stdout().flush();
    read_line().parse::<usize>().ok()
}

fn pause() {
    print!("{}  Press Enter to continue...{}", DIM, RESET);
    let _ = io::stdout().flush();
    read_line();
}

// ============================================
//  ACTIONS
// ============================================

fn action_add(list: &mut TodoList) {
    print!("{}  New task: {}", CYAN_B, RESET);
    let _ = io::stdout().flush();
    let title = read_line();

    match list.add(title) {
        Ok(id) => println!(
            "{}  ✓ {}Task [{}] added.",
            GREEN, RESET, id
        ),
        Err(e) => println!("{}  ✗ {}{}", RED, e, RESET),
    }
}

fn action_complete(list: &mut TodoList) {
    match read_id("Mark complete — Task ID:") {
        Some(id) => match list.complete(id) {
            Ok(_)  => println!(
                "{}  ✓ {}Task [{}] marked complete.",
                GREEN, RESET, id
            ),
            Err(e) => println!("{}  ✗ {}{}", RED, e, RESET),
        },
        None => println!("{}  ✗ Invalid ID.{}", RED, RESET),
    }
}

fn action_delete(list: &mut TodoList) {
    match read_id("Delete — Task ID:") {
        Some(id) => match list.delete(id) {
            Ok(_)  => println!(
                "{}  ✓ {}Task [{}] deleted.",
                GREEN, RESET, id
            ),
            Err(e) => println!("{}  ✗ {}{}", RED, e, RESET),
        },
        None => println!("{}  ✗ Invalid ID.{}", RED, RESET),
    }
}

fn action_clear(list: &mut TodoList) {
    let removed = list.clear_done();
    if removed == 0 {
        println!("{}  No completed tasks to clear.{}", DIM, RESET);
    } else {
        println!(
            "{}  ✓ {}Cleared {} completed task(s).",
            GREEN, RESET, removed
        );
    }
}

// ============================================
//  MAIN LOOP
// ============================================

fn main() {
    let mut list = TodoList::new();

    loop {
        print_header();
        render_tasks(&list);
        render_menu();

        let input = read_line();

        match input.as_str() {
            "1" => { action_add(&mut list);      pause(); }
            "2" => { action_complete(&mut list); pause(); }
            "3" => { action_delete(&mut list);   pause(); }
            "4" => { action_clear(&mut list);    pause(); }
            "0" => {
                clear_screen();
                println!(
                    "{}{}  Goodbye.\n{}",
                    BOLD, CYAN, RESET
                );
                break;
            }
            _ => {
                println!(
                    "{}  Unknown option.{}",
                    DIM, RESET
                );
                pause();
            }
        }
    }
}